### ISSUE TITLE
Multiple 'commented on' notifications about same post not getting grouped

### ISSUE THREAD

**Issue by balasankarc** [diaspora/diaspora#6623]
When same guy gave four comments on my post, I got 4 different notifications (See image)

![comment](https://cloud.githubusercontent.com/assets/2379271/12170250/9f1b6a60-b565-11e5-9ba7-6f5f53e547f8.png)

Pod Details (as per statistics.json)

Pod : Poddery
Version: 0.5.5.0-pe1956515

**Comment by jaywink** [comment]
I think this is the way it is supposed to work at the moment? Only different people get grouped, into unread notifications. Did you read the latest notification before the next one came or did it produce multiple unread notifications?

**Comment by balasankarc** [comment]
No. These are individual notifications about different comments. I thought grouping happened based on post.

**Comment by jaywink** [comment]
So new notifications got created even though you hadn't read the latest one? Sorry to repeat the question, not sure what the 'no' was for.

**Comment by balasankarc** [comment]
Oops. My bad.

New notifications got created **only after** I read all previous notifications.

I was under the belief, all notifications about a **post** got grouped, irrespective of whom it involved. But, now I think about it, that doesn't inform the user about multiple notifications from same person.

That being said, my notifications area full of notifications about only a single post is something I like to change, if possible. :)

**Comment by jaywink** [comment]
The question is "how". Since you had read the latest notification, the system, correctly, created a new notification for a new comment in the post. I guess old "identical" notifications could be cleaned from the list when creating a new version for a new comment, which would cut down entries in a long running discussion with one person for example.

**Comment by balasankarc** [comment]
Use the same notification, highlight it again in case there is a new notification about the same post? Multiple tabs opened would pose an issue though.

**Comment by jaywink** [comment]
Well, at least the timestamp would have to be refreshed, so essentially it is a new notification ;) Mind renaming the issue to this?

**Comment by balasankarc** [comment]
Is that what you suggested? If not, feel free to edit it (If you have edit permission. :) )

**Comment by SuperTux88** [comment]
And what if person A and B commented on a Post, you read the notification, and person B comments again? When you simply reuse the notification, you think that A **AND** B commented again. The only clean way is, how it is now: always create a new notification after you read the old.

**Comment by balasankarc** [comment]
Create a new notification per action per person per post . And reuse it.

**Comment by jaywink** [comment]
> And what if person A and B commented on a Post, you read the notification, and person B comments again? When you simply reuse the notification, you think that A AND B commented again. 

A + B != A or B

**Comment by SuperTux88** [comment]
ok, and what if you have this A+B notification, read it, B comments again => you create a new B notification.

So now: You don't mark this as read, A comments again too, what should you do?
* delete the B notification, and reuse the old A+B notification?
  * what happens when C comments then? change the old A+B notification and add C to it?
* add A to the new B notification?

Or:
> Create a new notification per action per person per post . And reuse it.

Do you want to disable the grouping completely? So when A+B comments, you create two separate notifications and if B comments again you can reuse the B notification?

**Comment by balasankarc** [comment]
So, it's gonna be a long list in either of the scenarios. Damn.

**Comment by SuperTux88** [comment]
as I already said: the only clean way to do this is how it is currently implemented: group the notifications (to reduce the notifications) and create new notifications (also grouped again) after you have read the old (to display clearly who has commented or liked your posts since you viewed the notifications last time).

**Comment by jaywink** [comment]
> ok, and what if you have this A+B notification, read it, B comments again => you create a new B notification.

What if we have a A + A + A + A + A + A + A comments on B's post. That is after all what this issue started as. It would be totally feasible to destroy the old previous versions of the comment.

This has absolutely no effect on the grouping that already works.

**Comment by SuperTux88** [comment]
A+A+A+A would be grouped together to a single A, but if you read it, a new A will be created. Or what if you have this: A comments, read A, A comments gain -> reuse old A, B comments? would you then group the B to the old reused A? What if you read this new A+B then, and A comments again? there will be no A to reuse, so create a new one, and it starts again ...

There is no consistent way to solve this problem, you make it worse when you change something how the current grouping works. A+A+A+A is a rare edge-case where it maybe make sense to reuse it, but there will always be a edge-case to this edge-case and B appears... so it doesn't work.

**Comment by jaywink** [comment]
Forget grouping A, just focus on "recycling" the latest single notification that A made, if it was read :D

> A+A+A+A is a rare edge-case where it maybe make sense to reuse it

Have a look at the image in the original issue description. Looks valid to me.

And forget about B. Writing logic to only "recycle" a notification if 1) the latest one was A only and it was read already is peanuts. There is no B, just A.

(ps. I think this minor issue is taking too much time from both of us - just see if someone comes up with a neat small patch ok?)

**Comment by SuperTux88** [comment]
ok, only if the last was from A and A comments again, then set a new date to the last one and mark it unread again? but you can't ignore B (at some point there will be a B commenting to the same post, thats a valid usecase), what if B comments then, would you then again group B to the reused A notification?

**Comment by davidmoseler** [comment]
I think you could do it according to the following rules:

-If there is no previous notification regarding that post, create a new one.
-If there is a notification telling that A1, A2, ..., Ak commented, and Ak+1 (a user not in the list) comments, then this notification (either read or unread) is updated to an unread notification telling that A1, A2, ..., Ak+1 commented.
-If there is a notification telling that A1, A2, ..., Ak commented, and Aj comments (j between 1 and k, that is, an user in the list), then this notification (either read or unread) is reused, going to the top of notifications, and it is made unread.

Do you think this could be a logical solution?

**Comment by davidmoseler** [comment]
I think you could do it according to the following rules:

-If there is no previous notification regarding that post, create a new one.
-If there is a notification telling that A1, A2, ..., Ak commented, and Ak+1 (a user not in the list) comments, then this notification (either read or unread) is updated to an unread notification telling that A1, A2, ..., Ak+1 commented.
-If there is a notification telling that A1, A2, ..., Ak commented, and Aj comments (j between 1 and k, that is, an user in the list), then this notification (either read or unread) is reused, going to the top of notifications, and it is made unread.

Do you think this could be a logical solution?
